monai/apps/pathology/transforms/post/dictionary.py

```diff
@@ -0,0 +1,152 @@
+    "PostProcessHoVerNetOutputD",
+    "PostProcessHoVerNetOutputd",
+]
+
+
+class GenerateSuccinctContourd(MapTransform):
+    """
+    Args:
+        height: height of bounding box, used to detect direction of line segment.
+        width: width of bounding box, used to detect direction of line segment.
+        allow_missing_keys: don't raise exception if key is missing.
+
+    """
+    backend = GenerateSuccinctContour.backend
+    def __init__(
+        self,
+        keys: KeysCollection,
+        height: int,
+        width: int,
+        allow_missing_keys: bool = False
+    ) -> None:
+        super().__init__(keys, allow_missing_keys)
+        self.converter = GenerateSuccinctContour(height=height, width=width)
+
+    def __call__(self, data):
+        d = dict(data)
+        for key in self.key_iterator(d):
+            d[key] = self.converter(d[key])
+
+        return d
+
+
+class PostProcessHoVerNetOutputd(MapTransform):
+    """
+    Dictionary-based transform that post processing image tiles. It assumes network has three branches, with a segmentation branch that
+    returns `np_pred`, a hover map branch that returns `hv_pred` and an optional classification branch that returns `nc_pred`. After this
+    tranform, it will return pixel-wise nuclear instance segmentation prediction and a instance-level information dictionary.
+
+    Args:
+        hv_pred_key: hover map branch output key. Defaults to `HoVerNet.Branch.HV.value`.
+        nv_pred_key: classification branch output key. Defaults to `HoVerNet.Branch.NC.value`.
+        inst_info_dict_key: a dict contaning a instance-level information dictionary will be added, which including bounding_box,
+            centroid and contour. If output_classes is not None, the dictionary will also contain pixel-wise nuclear type prediction.
+            Defaults to "inst_info".
+        output_classes: number of types considered at output of NC branch.
+        return_centroids: whether to generate coords for each nucleus instance.
+            Defaults to True.
+        threshold_pred: threshold the float values of prediction to int 0 or 1 with specified theashold. Defaults to 0.5.
+        threshold_overall: threshold the float values of overall gradient map to int 0 or 1 with specified theashold.
+            Defaults to 0.4.
+        min_size: objects smaller than this size are removed. Defaults to 10.
+        sigma: std. could be a single value, or `spatial_dims` number of values. Defaults to 0.4.
+        kernel_size: the size of the Sobel kernel. Defaults to 17.
+        radius: the radius of the disk-shaped footprint. Defaults to 2.
+        allow_missing_keys: don't raise exception if key is missing.
+    """
+
+    backend = PostProcessHoVerNetOutput.backend
+
+    def __init__(
+        self,
+        keys: KeysCollection = HoVerNet.Branch.NP.value,
+        hv_pred_key: str = HoVerNet.Branch.HV.value,
+        nv_pred_key: str = HoVerNet.Branch.NC.value,
```

Comment:
Hi @wyli @ericspod @rijobro @Nic-Ma,

I would like to know your opinion on what is the best way to create map transforms for transforms that rely on multiple keys at the same time to do the transforms. Here is an example where the output of HoVerNet model is a dictionary with three keys (one for each of three branches). This post-processing transform need to accept three keys but our `MapTransform` only accepts one set of independent keys `keys`.

This also has been the case in deep grow transform, where the dictionary transform inherit `Transform` instead of `MapTransform` and create its own signature.
https://github.com/Project-MONAI/MONAI/blob/ba670c9ba78f9f4d76eb12b4d33f245b5774d462/monai/apps/deepgrow/transforms.py#L79-L81

It seems that it is not a one-off thing and may happen more, thus I believe it worth to discuss and come to an agreement on a "right way" to do that in MONAI. What do you thing?

Reply:
This seems fine to me. `keys` can be a sequence of keys on which the transform will be applied. Other references to keys (could be single keys as in this case, could feasibly be sequence of keys (e.g., meta_key prior to us implementing MetaTensor) will be used as supporting information. `CropForegroundd` ([link](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/croppad/dictionary.py#L621)), for example, will crop all images given as `keys`, but will use the `source_key` as the basis for deciding what constitutes foreground and background.

This seems to be the way we do it throughout, do you have any examples of us doing it differently? Or suggestions on how to improve upon it?

Reply:
Thanks @rijobro! The situations that you mentioned make sense and it seems that `keys` argument are the primary keys and the other keys are kind of auxiliary or complementary.

Our situation is that all the keys have equal value, so even the decision of which one should be `keys` and which one `other_name_keys` is arbitrary. This might cause a confusion.

The other way that we have done in MONAI (and it might be better for these situations) is in deepgrow transforms, where we have inherited `Transform` instead of `MapTransform` and made equally important key arguments with different names, e.g., [AddInitialSeedPointd](https://github.com/Project-MONAI/MONAI/blob/ba670c9ba78f9f4d76eb12b4d33f245b5774d462/monai/apps/deepgrow/transforms.py#L68).

Reply:
Ah I see. In the [proposed](https://github.com/Project-MONAI/MONAI/pull/5186/files/48f796f8e7bd49b3ddbb606c24332b1298e42071#diff-980b604c7a16c6f010fea135e006ad1f7b65a9744a311df9421e9feb74e78b10R145) implementation of `PostProcessHoVerNetOutputd`, even though different keys might be equally important in the calculation, the value in `[key]` is the one that is changed, so I think the way it's currently suggested is correct. You also update a dictionary info key, but I would argue that it's still secondary information compared to the image that is being replaced with `[key]`.

Reply:
Hi @wyli @ericspod ,

What do you guys think about this question?

Thanks in advance.

Reply:
@rijobro you are right, in the current implementation the key in `keys` is the one that is changed but it can be either of branches (NP or HV), or better not to be any of them and create a new key.

```python
        keys: KeysCollection = HoVerNet.Branch.NP.value,
        hv_pred_key: str = HoVerNet.Branch.HV.value,
```

Reply:
Ah, I see...

Reply:
Hi @drbeh, in my mind, in our current case, the NP branch should be the one that is changed, as HV branch is actually used to assist in generating the instance-level segmentation. NP branch's output is indeed the segmentation result. Thanks!

Reply:
@KumoLiu you are right but actually none of them is instance segmentation. Even if it was the case, the approach that we took in deepgrow is more readable and less confusing for the user. You just need to inherit from Transform and give the argument for each key the appropriate name (e.g., `np_pred_key` and `hv_pred_key` as you did). What do you think?
https://github.com/Project-MONAI/MONAI/blob/ba670c9ba78f9f4d76eb12b4d33f245b5774d462/monai/apps/deepgrow/transforms.py#L87-L90